output/cloud/expv2/output_test.go

```diff
@@ -114,3 +115,163 @@ func TestOutputCollectSamples(t *testing.T) {
+		},
+		{
+			name: "error code equals 4 but no abort",
+			err: cloudapi.ErrorResponse{
+				Response: &http.Response{StatusCode: http.StatusForbidden},
+				Code:     4,
+			},
+			abort:                   false,
+			expStopMetricCollection: true,
+			expAborted:              false,
+		},
+		{
+			name: "error code equals 4 and abort",
+			err: cloudapi.ErrorResponse{
+				Response: &http.Response{StatusCode: http.StatusForbidden},
+				Code:     4,
+			},
+			abort:                   true,
+			expStopMetricCollection: true,
+			expAborted:              true,
+		},
+	}
+
+	for _, tc := range tests {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			stopFuncCalled := false
+			stopMetricCollection := false
+
+			o := Output{
+				logger: testutils.NewLogger(t),
+				testStopFunc: func(error) {
+					stopFuncCalled = true
+				},
+				abort: make(chan struct{}),
+			}
+			o.config.StopOnError = null.BoolFrom(tc.abort)
+
+			done := make(chan struct{})
+			stopped := make(chan struct{})
+
+			go func() {
+				defer close(stopped)
+
+				<-done
+				select {
+				case <-o.abort:
+					stopMetricCollection = true
+				default:
+				}
+			}()
+
+			o.handleFlushError(tc.err)
+			close(done)
+			<-stopped
+
+			assert.Equal(t, tc.expStopMetricCollection, stopMetricCollection)
+			assert.Equal(t, tc.expAborted, stopFuncCalled)
+		})
+	}
+}
+
+// assert that the output does not stuck or panic
+// when it is called doesn't stuck
+func TestOutputHandleFlushErrorMultipleTimes(t *testing.T) {
+	t.Parallel()
+
+	var stopFuncCalled int
+	o := Output{
+		logger: testutils.NewLogger(t),
+		abort:  make(chan struct{}),
+		testStopFunc: func(error) {
+			stopFuncCalled++
+		},
+	}
+	o.config.StopOnError = null.BoolFrom(true)
+
+	er := cloudapi.ErrorResponse{
+		Response: &http.Response{
+			StatusCode: http.StatusForbidden,
+		},
+		Code: 4,
+	}
+	o.handleFlushError(fmt.Errorf("first error: %w", er))
+	o.handleFlushError(fmt.Errorf("second error: %w", er))
+	assert.Equal(t, 1, stopFuncCalled)
+}
+
+func TestOutputAddMetricSamples(t *testing.T) {
+	t.Parallel()
+
+	stopSamples := make(chan struct{})
+	o := Output{
+		abort: stopSamples,
+	}
+	require.Empty(t, o.GetBufferedSamples())
+
+	s := metrics.Sample{}
+	o.AddMetricSamples([]metrics.SampleContainer{
+		metrics.Samples([]metrics.Sample{s}),
+	})
+	require.Len(t, o.GetBufferedSamples(), 1)
+
+	// Not accept samples anymore when the chan is closed
+	close(stopSamples)
+	o.AddMetricSamples([]metrics.SampleContainer{
+		metrics.Samples([]metrics.Sample{s}),
+	})
+	require.Empty(t, o.GetBufferedSamples())
+}
+
+func TestOutputPeriodicInvoke(t *testing.T) {
+	t.Parallel()
+
+	stop := make(chan struct{})
+	var called uint64
+	cb := func() {
+		updated := atomic.AddUint64(&called, 1)
+		if updated == 2 {
+			close(stop)
+		}
+	}
+	o := Output{stop: stop}
+	o.periodicInvoke(time.Duration(1), cb) // loop
+	<-stop
+	assert.Greater(t, atomic.LoadUint64(&called), uint64(1))
```

Comment:
Why not `assert.Equal()` here?

Reply:
@imiric Because there isn't a guarantee that the close chan is picked before doing a new iteration with a so small time frame.

I don't know if there is a time frame that would guarantee us a sync flow but it would mean relying on time.